### PR TITLE
Add Eslint and fix errors

### LINKED ISSUE #1: Setup ESLint
Task:
- Add ESLint to project
- Use ESLint plugins - `eslint-plugin-react`, `eslint-config-airbnb`
- Fix ESLint issues

Resources:
- https://create-react-app.dev/docs/setting-up-your-editor/#extending-or-replacing-the-default-eslint-config
- https://www.npmjs.com/package/eslint-plugin-react
- https://www.npmjs.com/package/eslint-config-airbnb

### PR DESCRIPTION
**Pull request by ekostandov**
Fixed ESlint problems.

### CODE REVIEW

File: .eslintignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+/node_modules
+/public
+/src/registerServiceWorker.js
```
Comment (singvarr) [inline .eslintignore]: Remove it and add a new line at the end

File: src/App/App.js [hunk 2]
```diff
@@ -1,20 +1,16 @@
-import React from 'react'
+import '../common/style/reset.css';
+import '../common/style/base.css';
 
-import "../common/style/reset.css"
-import "../common/style/base.css"
+import Header from './Header/Header.js';
```
Comment (singvarr) [inline src/App/App.js]: Remove extension

File: src/App/Header/Cart.js [hunk 3]
```diff
@@ -1,41 +1,29 @@
...
+import { Link } from 'react-router-dom';
+import { connect } from 'react-redux';
 
-import "./Cart.css"
+import './Cart.css';
 
-import products, { getProductsMap } from '../Main/Product/products'
-import CartTotal from '../../Features/Cart/CartTotal'
-import CartProductList from '../../Features/Cart/CartProductList'
+//import products, { getProductsMap } from '../Main/Product/products';
```
Comment (singvarr) [inline src/App/Header/Cart.js]: Remove commented code

File: src/App/Main/Cart/CartPage.js [hunk 4]
```diff
@@ -1,46 +1,43 @@
...
-import CartProductList from '../../../Features/Cart/CartProductList'
-import './CartPage.css'
+import { Link } from 'react-router-dom';
+import PropTypes from 'prop-types';
+import CartTotal from '../../../Features/Cart/CartTotal';
+import CartProductList from '../../../Features/Cart/CartProductList';
+import './CartPage.css';
 import CartProductListItemExtended from '../../../Features/Cart/CartProductListItemExtended';
-import { connect } from 'react-redux'
+import { connect } from 'react-redux';
```
Comment (singvarr) [inline src/App/Main/Cart/CartPage.js]: npm dependencies should be placed before our modules

File: src/App/Main/Product/ProductList.js [hunk 5]
```diff
@@ -1,42 +1,43 @@
-import React from 'react'
+import products from './products.js';
```
Comment (singvarr) [inline src/App/Main/Product/ProductList.js]: Remove extensions

File: src/App/Main/Product/ProductList.js [hunk 6]
```diff
@@ -1,42 +1,43 @@
...
+import ProductListItem from './ProductListItem.js';
+import PropTypes from 'prop-types';
```
Comment (singvarr) [inline src/App/Main/Product/ProductList.js:1]: npm dependencies should be before

File: src/App/Main/Product/ProductListItem.js [hunk 7]
```diff
@@ -1,136 +1,139 @@
...
-import PropTypes from 'prop-types'
-import './ProductListItem.css'
-import QuantityInput from '../../../Component/QuantityInput/QuantityInput'
-import { connect } from 'react-redux'
-import { Link } from 'react-router-dom'
+import { Component } from 'react';
+import PropTypes from 'prop-types';
+import './ProductListItem.css';
+import QuantityInput from '../../../Component/QuantityInput/QuantityInput';
+import { connect } from 'react-redux';
```
Comment (singvarr) [inline src/App/Main/Product/ProductListItem.js:3]: fix dependencies order

File: src/App/Main/Product/ProductListItem.js [hunk 8]
```diff
@@ -1,136 +1,139 @@
...
+      description = 'bla bla bla..',
+      image,
+      type,
+      screenSize,
+      capacity,
+      price,
+      dispatchAddToCartButton,
+      isLiked,
+    } = this.props;
+    // == const name = this.props.name;
```
Comment (singvarr) [inline src/App/Main/Product/ProductListItem.js]: remove commented code

File: src/App/Main/Product/ProductListItem.js [hunk 9]
```diff
@@ -1,136 +1,139 @@
...
+    dispatch({
+      type: 'ADD_PRODUCT_TO_CART',
+      id,
+      count,
+    }),
+});
 
+export default connect(mapStateToProps, mapDispatchToProps)(ProductListItem);
 
 // const ProductListItem = ({
```
Comment (singvarr) [inline src/App/Main/Product/ProductListItem.js]: remove comment

File: src/App/Main/Product/ProductPage.js [hunk 10]
```diff
@@ -1,13 +1,14 @@
...
+
+import products, { getProductsMap } from './products';
+
+const ProductPage = ({ productItem = getProductsMap(products), match }) => (
+  <div>{productItem[match.params.productId].name}</div>
+);
+
+ProductPage.propTypes = {
+  productItem: PropTypes.array,
+  match: PropTypes.object,
```
Comment (singvarr) [inline src/App/Main/Product/ProductPage.js]: describe fields of `match`. Use `PropTypes.shape`

File: src/Features/Cart/CartProductListItemExtended.js [hunk 11]
```diff
@@ -1,83 +1,93 @@
-import React from 'react'
-import './CartProductListItemExtended.css'
+import './CartProductListItemExtended.css';
+import PropTypes from 'prop-types';
```
Comment (singvarr) [inline src/Features/Cart/CartProductListItemExtended.js:2]: fix order

File: src/index.js [hunk 12]
```diff
@@ -1,22 +1,21 @@
...
-    document.getElementById('root'));
-registerServiceWorker();
+  <Provider store={store}>
+    <BrowserRouter>
+      <App />
+    </BrowserRouter>
+  </Provider>,
+  document.getElementById('root')
+);
+// registerServiceWorker();
```
Comment (singvarr) [inline src/index.js]: remove comment

**Review by singvarr — CHANGES_REQUESTED** [review]
Add these rules to ESLint:
- Forbid file extension if file is `.js` or `.json`
- Add a rule that will control the order of imports: 1. npm-modules 2. modules from parent directories 3. modules in the same folder. See `eslint-plugin-import`

File: src/App/App.js [hunk 13]
```diff
@@ -1,20 +1,16 @@
-import React from 'react'
+import '../common/style/reset.css';
```
Comment (singvarr) [inline src/App/App.js]: CSS imports should be after JS modules

File: src/App/Main/Cart/CartPage.js [hunk 14]
```diff
@@ -1,46 +1,45 @@
...
-const mapStateToProps = () => (
-    (state) => {
-        return {
-            productsInCart: state.productsInCart,
-        }
-    }
-)
+CartPage.propTypes = {
+  productsInCart: PropTypes.array,
+  // productMap: PropTypes.array,
```
Comment (singvarr) [inline src/App/Main/Cart/CartPage.js]: Remove comment

File: src/App/Main/Product/ProductListItem.js [hunk 15]
```diff
@@ -1,181 +1,137 @@
...
-            this.props.dispatchDislike(this.props.id)
-        } else {
-            this.props.dispatchLike(this.props.id)
-        }
+class ProductListItem extends Component {
+  static propTypes = {
+    name: PropTypes.string.isRequired,
+    description: PropTypes.string,
+    image: PropTypes.string,
+    type: PropTypes.string.isRequired,
```
Comment (singvarr) [inline src/App/Main/Product/ProductListItem.js:15]: If prop is optional add default value of prop in `defaultProps`. Add ESLint rule that will validate it

**Review by singvarr — CHANGES_REQUESTED** [review]
Fix these errors everywhere in the app

File: .eslintrc.json [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+    "import/extensions": ["error", { "js": "never", "json": "never" }],
+    "import/order": [
+      "error",
+      {
+        "groups": ["external", "parent", "sibling"]
+      }
+    ],
+    "react/require-default-props": [
+      "error",
+      { "forbidDefaultForRequired": false, "ignoreFunctionalComponents": false }
```
Comment (singvarr) [inline .eslintrc.json]: Make `forbidDefaultForRequired` `true`

File: src/App/Main/Product/ProductListItem.js [hunk 17]
```diff
@@ -1,181 +1,143 @@
...
+    description: PropTypes.string,
+    image: PropTypes.string.isRequired,
+    type: PropTypes.string.isRequired,
+    screenSize: PropTypes.number,
+    capacity: PropTypes.number,
+    price: PropTypes.number.isRequired,
+    isLiked: PropTypes.bool,
+    dispatchDislike: PropTypes.func,
+    id: PropTypes.number.isRequired,
+    dispatchLike: PropTypes.func,
```
Comment (singvarr) [inline src/App/Main/Product/ProductListItem.js]: Not typesafe. If it's optional you cannot call it without extra check (e.g. `if (dispatchLike) dispatchLike();`). In this case, all handlers from the store should be required.

File: src/App/Main/Product/ProductListItem.js [hunk 18]
```diff
@@ -1,181 +1,143 @@
...
-                className="btn-add-to-cart"
-                onClick={ () => dispatchAddToCartButton(id,this.state.productCount)}
-            >Add to cart</button>
+  };
+
+  render() {
+    const {
+      id,
+      name,
+      description = 'bla bla bla..',
```
Comment (singvarr) [inline src/App/Main/Product/ProductListItem.js]: Remove it. `defaultProps` will handle it

File: src/App/Main/Product/ProductPage.js [hunk 19]
```diff
@@ -1,13 +1,27 @@
...
+
+ProductPage.propTypes = {
+  productItem: PropTypes.arrayOf(
+    PropTypes.shape({
+      name: PropTypes.string,
+    })
+  ),
+  match: PropTypes.shape({
+    params: PropTypes.shape({
+      productId: PropTypes.number,
```
Comment (singvarr) [inline src/App/Main/Product/ProductPage.js]: Should be required

File: src/Component/QuantityInput/QuantityInput.js [hunk 20]
```diff
@@ -1,26 +1,37 @@
...
-            disabled={quantity > maxValue}
-        >+
-        </button>
-    </span>
-)
+  <span className='product-quantity'>
+    <button onClick={() => onDecrementClick()} disabled={quantity <= minValue}>
+      -
+    </button>
+    <input type='text' value={quantity} readOnly />
```
Comment (singvarr) [inline src/Component/QuantityInput/QuantityInput.js]: Remove `type='text'`. It's default type

File: src/Component/QuantityInput/QuantityInput.js [hunk 21]
```diff
@@ -1,26 +1,37 @@
...
+      +
+    </button>
+  </span>
+);
 
-export default QuantityInput
\ No newline at end of file
+QuantityInput.propTypes = {
+  quantity: PropTypes.number,
+  onDecrementClick: PropTypes.func,
```
Comment (singvarr) [inline src/Component/QuantityInput/QuantityInput.js]: Not typesafe. Make it required or add an extra check.
If handler is absent, make `null` as fallback value

File: src/Features/Cart/CartProductListItem.js [hunk 22]
```diff
@@ -1,14 +1,21 @@
...
+const CartProductListItem = ({ product, productCount }) => (
+  <li>
+    <span> {product.name} : </span>
+    <span> {productCount} : </span>
+    <span> {product.price * productCount} </span>
+  </li>
+);
+
+CartProductListItem.propTypes = {
+  product: PropTypes.object,
```
Comment (singvarr) [inline src/Features/Cart/CartProductListItem.js]: TODO: check is required

File: src/Features/Cart/CartTotal.js [hunk 23]
```diff
@@ -1,19 +1,30 @@
...
-        {
-            keys(productsInCart).reduce((total,productId) => {
-                return total + (productMap[productId].price * productsInCart[productId])
-            },0)
-        }
-    </div>
-)
+  <div>
+    Total: $
+    {keys(productsInCart).reduce(
```
Comment (singvarr) [inline src/Features/Cart/CartTotal.js]: 1. Use `Object.values` and remove `lodash` import
2. Move this operation to constant above

File: src/Features/Cart/CartTotal.js [hunk 24]
```diff
@@ -1,19 +1,30 @@
...
+import { keys } from 'lodash';
+import PropTypes from 'prop-types';
+
+import products, { getProductsMap } from '../../App/Main/Product/products';
 
 const CartTotal = ({
-    productsInCart,
-    productMap = getProductsMap(products),
+  productsInCart,
+  productMap = getProductsMap(products),
```
Comment (singvarr) [inline src/Features/Cart/CartTotal.js]: Use `defaultProps` for fallback

File: src/Features/Cart/cart.reducer.js [hunk 25]
```diff
@@ -1,30 +1,35 @@
...
+    case 'SET_PRODUCT_QUANTITY_IN_CART': {
+      return {
+        ...state,
+        [action.id]: action.quantity,
+      };
+    }
+
+    case 'REMOVE_PRODUCT_FROM_CART': {
+      return omit(state, action.id);
+    } // delete id from state
```
Comment (singvarr) [inline src/Features/Cart/cart.reducer.js]: Remove comment

File: src/Features/Cart/CartProductList.js [hunk 26]
```diff
@@ -1,30 +1,45 @@
...
 
 const CartProductList = ({
-    productsInCart,
-    productMap = getProductsMap(products),
-    ListItemComponent = CartProductListItem,
-    removeProductFromCart,
-    setProductQuantityInCart,
+  productsInCart,
+  productMap = getProductsMap(products),
+  ListItemComponent = CartProductListItem,
```
Comment (singvarr) [inline src/Features/Cart/CartProductList.js]: Do we need to pass this component via props? Can we just import `CartProductListItem`

File: src/Features/Cart/CartProductList.js [hunk 27]
```diff
@@ -1,30 +1,45 @@
...
-                            setProductQuantityInCart={setProductQuantityInCart}
-                        />
-                    ))
-                }
-        </ul>
-    </div>
-)
+  <div className='cart-product-list'>
+    <ul>
+      {keys(productsInCart).map(productId => (
```
Comment (singvarr) [inline src/Features/Cart/CartProductList.js]: Use `Object.entries`

File: src/App/Main/Product/ProductListItem.js [hunk 28]
```diff
@@ -1,181 +1,143 @@
...
+        onDecrementClick={this.onDecrementClick}
+        onIncrementClick={this.onIncrementClick}
+        minValue={1}
+        maxValue={9}
+      />
+    );
+  }
+
+  onLikeClick = () => {
+    if (this.props.isLiked) {
```
Comment (singvarr) [inline src/App/Main/Product/ProductListItem.js]: Destructure props

**Review by singvarr — CHANGES_REQUESTED** [review]

File: src/App/Main/Product/ProductListItem.js (hunk 17 above)
Comment (ekostandov) [reply src/App/Main/Product/ProductListItem.js]: I temporary made functions not required and set default value to null.

File: src/Features/Cart/CartTotal.js (hunk 23 above)
Comment (ekostandov) [reply src/Features/Cart/CartTotal.js]: Done, but haven't found the way how to move this operation out of the component.

File: src/App/Main/Product/ProductListItem.js (hunk 28 above)
Comment (ekostandov) [reply src/App/Main/Product/ProductListItem.js]: Didn't catch it. :(

**Pull request merged**